File: spark-on-angel/mllib/src/main/java/com/tencent/angel/spark/ml/psf/embedding/line/Dot.scala
/*
 * Tencent is pleased to support the open source community by making Angel available.
 *
 * Copyright (C) 2017-2018 THL A29 Limited, a Tencent company. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in 
 * compliance with the License. You may obtain a copy of the License at
 *
 * https://opensource.org/licenses/Apache-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 *
 */


package com.tencent.angel.spark.ml.psf.embedding.line

import java.text.SimpleDateFormat
import java.util.Date
import scala.collection.JavaConversions._
import scala.util.Random

import io.netty.buffer.ByteBuf
import org.apache.commons.logging.{Log, LogFactory}

import com.tencent.angel.PartitionKey
import com.tencent.angel.ml.matrix.psf.get.base.{GetParam, PartitionGetParam}
import com.tencent.angel.ps.storage.matrix.ServerPartition
import com.tencent.angel.psagent.PSAgentContext
import com.tencent.angel.spark.ml.psf.embedding.line.Dot.{DotParam, DotPartitionParam}
import com.tencent.angel.spark.ml.psf.embedding.{NEDot, NENegativeSample, PartitionWrapper}

class Dot(param: GetParam) extends NEDot(param) {

  val log: Log = LogFactory.getLog(this.getClass)

  def this(matrixId: Int,
           src: Array[Int],
           dst: Array[Int],
           seed: Int,
           numNegSample: Int,
           maxIndex: Int,
           parDim: Int,
           order: Int) =
    this(new DotParam(matrixId, src, dst, seed, numNegSample, maxIndex, parDim, order))

  def this() = this(null)

  override def doProcess(part: ServerPartition, pParam: PartitionGetParam): Array[Float] = {
    val partParam = pParam.asInstanceOf[DotPartitionParam]
    assert(partParam.src.length == partParam.dst.length)
    val pw = new PartitionWrapper(part, partParam.partDim, partParam.order)
    val src = partParam.src
    val dst = partParam.dst
    val numNegSample = partParam.numNegSample
    val sg = new NENegativeSample(partParam.maxIndex, partParam.seed)

    val startTime = System.currentTimeMillis()
    val dotRet = new Array[Float](src.length * 2 * (1 + numNegSample))
    var pos = 0
    for (i <- 0 until src.length * 2) {
      val (srcIdx, dstIdx) = if (i % 2 == 0) (src(i / 2), dst(i / 2)) else (dst(i / 2), src(i / 2))
      dotRet(pos) = pw.dot(srcIdx, dstIdx)
      pos += 1
      for (_ <- 0 until numNegSample) {
        dotRet(pos) = pw.dot(srcIdx, sg.next(srcIdx))
        pos += 1
      }
    }
    if (Random.nextDouble() < 0.001) logTime(startTime)
    dotRet
  }

  private def logTime(startTime: Long): Unit = {
    val start = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss").format(new Date(startTime))
    val cost = System.currentTimeMillis() - startTime
    log.info(s"dot start in $start, time cost: ${cost}ms")
  }
}

object Dot {

  class DotPartitionParam(matrixId: Int,
                          partKey: PartitionKey,
                          var src: Array[Int],
                          var dst: Array[Int],
                          var seed: Int,
                          var numNegSample: Int,
                          var maxIndex: Int,
                          var partDim: Int,
                          var order: Int) extends PartitionGetParam(matrixId, partKey) {
    def this() = this(-1, null, null, null, -1, -1, -1, -1, -1)

    override def serialize(buf: ByteBuf): Unit = {
      super.serialize(buf)
      buf.writeInt(src.length)
      src.indices.foreach { i =>
        buf.writeInt(src(i))
        buf.writeInt(dst(i))
      }
      buf.writeInt(seed)
      buf.writeInt(numNegSample)
      buf.writeInt(maxIndex)
      buf.writeInt(partDim)
      buf.writeInt(order)
    }

    override def deserialize(buf: ByteBuf): Unit = {
      super.deserialize(buf)
      val length = buf.readInt
      this.src = new Array[Int](length)
      this.dst = new Array[Int](length)
      src.indices.foreach { i =>
        this.src(i) = buf.readInt()
        this.dst(i) = buf.readInt()
      }
      this.seed = buf.readInt
      this.numNegSample = buf.readInt
      this.maxIndex = buf.readInt
      this.partDim = buf.readInt
      this.order = buf.readInt
    }

    override def bufferLen: Int = super.bufferLen + 8 * src.length + 24
  }

  class DotParam(matrixId: Int,
                 var src: Array[Int],
                 var dst: Array[Int],
                 var seed: Int,
                 var numNegSample: Int,
                 var maxIndex: Int,
                 var vecNumPreNum: Int,
                 var order: Int) extends GetParam(matrixId) {

    override def split: java.util.List[PartitionGetParam] = {
      PSAgentContext.get.getMatrixMetaManager.getPartitions(matrixId)
        .map(new DotPartitionParam(matrixId, _, src, dst, seed, numNegSample, maxIndex, vecNumPreNum, order))
    }
  }

}

File: spark-on-angel/mllib/src/main/java/com/tencent/angel/spark/ml/psf/embedding/line/Dot.java
/*
 * Tencent is pleased to support the open source community by making Angel available.
 *
 * Copyright (C) 2017-2018 THL A29 Limited, a Tencent company. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in
 * compliance with the License. You may obtain a copy of the License at
 *
 * https://opensource.org/licenses/Apache-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 *
 */

package com.tencent.angel.spark.ml.psf.embedding.line;

import com.tencent.angel.PartitionKey;
import com.tencent.angel.exception.AngelException;
import com.tencent.angel.ml.math2.storage.IntFloatDenseVectorStorage;
import com.tencent.angel.ml.matrix.psf.get.base.GetFunc;
import com.tencent.angel.ml.matrix.psf.get.base.GetResult;
import com.tencent.angel.ml.matrix.psf.get.base.PartitionGetParam;
import com.tencent.angel.ml.matrix.psf.get.base.PartitionGetResult;
import com.tencent.angel.ps.storage.vector.ServerIntFloatRow;
import com.tencent.angel.ps.storage.vector.ServerRowUtils;
import com.tencent.angel.spark.ml.psf.embedding.NEDot;
import com.tencent.angel.spark.ml.psf.embedding.ServerWrapper;
import io.netty.buffer.ByteBuf;

import java.util.List;

public class Dot extends GetFunc {

  public Dot(DotParam param) {
    super(param);
  }

  public Dot() {
    super(null);
  }

  @Override
  public PartitionGetResult partitionGet(PartitionGetParam partParam) {
    if (partParam instanceof DotPartitionParam) {

      DotPartitionParam param = (DotPartitionParam) partParam;

      float[] dots;
      try {
        // some params
        PartitionKey pkey = param.getPartKey();

        int seed = param.seed;
        ByteBuf edges = param.edgeBuf;

        // max index for node/word
        int maxIndex = ServerWrapper.getMaxIndex();
        int negative = ServerWrapper.getNegative();
        int partDim = ServerWrapper.getPartDim();
        int order = ServerWrapper.getOrder();

        // compute number of nodes for one row
        int size = (int) (pkey.getEndCol() - pkey.getStartCol());
        int numNodes = size / (partDim * order);

        int numRows = pkey.getEndRow() - pkey.getStartRow();
        float[][] layers = new float[numRows][];
        for (int row = 0; row < numRows; row++)
          layers[row] = ServerRowUtils.getVector((ServerIntFloatRow) psContext.getMatrixStorageManager()
              .getRow(pkey, row)).getStorage()
              .getValues();

        EmbeddingModel model;
        switch (order) {
          case 1:
            model = new LINEFirstOrderModel(partDim, negative, seed, maxIndex, numNodes, layers);
            break;
          default:
            model = new LINESecondOrderModel(partDim, negative, seed, maxIndex, numNodes, layers);
        }
        dots = model.dot(edges);
        ServerWrapper.setNumInputs(param.partitionId, model.getNumInputsToUpdate());
        ServerWrapper.setNumOutputs(param.partitionId, model.getNumOutputsToUpdate());
      } finally {
        param.clear();
      }

      return new DotPartitionResult(dots);
    }

    return null;
  }

  @Override
  public GetResult merge(List<PartitionGetResult> partResults) {
    if (partResults.size() > 0 && partResults.get(0) instanceof DotPartitionResult) {
      int size = ((DotPartitionResult) partResults.get(0)).length;

      // check the length of dot values
      for (PartitionGetResult result : partResults) {
        if (result instanceof DotPartitionResult &&
            size != ((DotPartitionResult) result).length)
          throw new AngelException(
              String.format("length of dot values not same one is %d other is %d",
                  size,
                  ((DotPartitionResult) result).length));
      }

      // merge dot values from all partitions
      float[] results = new float[size];
      for (PartitionGetResult result : partResults)
        if (result instanceof DotPartitionResult)
          try {
            ((DotPartitionResult) result).merge(results);
          } finally {
            ((DotPartitionResult) result).clear();
          }
      return new NEDot.NEDotResult(results);
    }

    return null;
  }
}
